File: ml-benchmark/synduce/constraints/ensures/mps_no_ensures.f
Config ExtraGrammar = "AutoLifter";
Inductive CList = cnil Unit | single Int | concat {CList, CList};
Inductive List = nil Unit | cons {Int, List};

cat = fix (\f: List->List->List. \x: List. \y: List.
  match x with
    cons {h, t} -> cons {h, f t y}
  | nil _ -> y
  end
);

repr = fix (
  \f: CList -> Compress List. \cl: CList.
  match cl with
    cnil _ -> nil unit
  | single h -> cons {h, nil unit}
  | concat {l, r} -> cat (f l) (f r)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

spec = \xs: List. (fix (
  \f: List -> {Int, Int}. \xs: List.
  match xs with
    nil _ -> {0, 0}
  | cons {h, t} ->
    let res = f t in
      {max 0 (+ res.1 h), + res.2 h}
  end
) xs).1;

@Start main = \cl: CList. spec (repr cl);

File: ml-benchmark/synduce/constraints/ensures/mps_no_ensures.ml
config EnableDeepCoder = true

type clist = Cnil of unit | Single of int | Concat of clist * clist
type ilist = Nil of unit | Cons of int * ilist

let rec cat x y =
  match x with
  | Cons (h, t) -> Cons (h, cat t y)
  | Nil _ -> y

val repr: clist -> ilist compress
let rec repr cl =
  match cl with
  | Cnil _ -> Nil ()
  | Single h -> Cons (h, Nil ())
  | Concat (l, r) -> cat (repr l) (repr r)

let max a b = if a < b then b else a

let spec xs =
  let rec f xs =
    match xs with
    | Nil _ -> (0, 0)
    | Cons (h, t) ->
        let res = f t in
        match res with
        | (a1, a2) -> (max 0 (a1 + h), a2 + h)
  in
  let tmp = f xs in
  match tmp with
  | (v, _) -> v

let program cl = spec (repr cl)